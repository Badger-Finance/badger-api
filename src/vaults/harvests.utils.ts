import { greaterThanOrEqualTo } from '@aws/dynamodb-expressions';
import {
  formatBalance,
  gqlGenT,
  keyBy,
  Vault__factory,
  VaultHarvestData,
  YieldEvent,
  YieldType,
} from '@badger-dao/sdk';
import { BadgerTreeDistribution_OrderBy, SettHarvest_OrderBy } from '@badger-dao/sdk/lib/graphql/generated/badger';
import { BigNumber, ethers } from 'ethers';

import { getDataMapper, getVaultEntityId } from '../aws/dynamodb.utils';
import { VaultDefinitionModel } from '../aws/models/vault-definition.model';
import { VaultYieldEvent } from '../aws/models/vault-yield-event.model';
import { Chain } from '../chains/config/chain.config';
import { OrderDirection } from '../graphql/generated/balancer';
import { queryPriceAtTimestamp } from '../prices/prices.utils';
import { getFullToken } from '../tokens/tokens.utils';
import { getInfuelnceVaultYieldBalance, isInfluenceVault } from './influence.utils';
import { VaultYieldItem } from './interfaces/vault-yield-item.interface';
import { VAULT_TWAY_DURATION } from './vaults.config';
import { getStrategyInfo } from './vaults.utils';
import { calculateYield } from './yields.utils';

// this allows five chunks of 10k blocks per index, approx 2 week
export const HARVEST_SCAN_BLOCK_INCREMENT = 100_000;

function grabYieldGraphTx(graphTxId?: string): string {
  if (!graphTxId) {
    return '';
  }

  if (!graphTxId.includes('-')) {
    return graphTxId;
  }

  return graphTxId.split('-')[0];
}

/**
 * Modify subgraph response to match on chain event data allowing it to fit into our estimation functions.
 * @param vault vault requesting graph data transformation
 * @param harvests harvests data retrieved from the graph
 * @param distributions distribution data retrieved from the graph
 * @returns vault harvest data in the same form as delivered via event logs
 */
function constructGraphVaultData(
  vault: VaultDefinitionModel,
  harvests: gqlGenT.SettHarvestsQuery['settHarvests'],
  distributions: gqlGenT.BadgerTreeDistributionsQuery['badgerTreeDistributions'],
): VaultHarvestData[] {
  const harvestsByTimestamp = keyBy(harvests, (harvest) => harvest.timestamp);
  const treeDistributionsByTimestamp = keyBy(distributions, (distribution) => distribution.timestamp);
  const timestamps = Array.from(
    new Set([...harvestsByTimestamp.keys(), ...treeDistributionsByTimestamp.keys()]).values(),
  );
  return timestamps.map((t) => {
    const timestamp = Number(t);
    const currentHarvests = harvestsByTimestamp.get(timestamp) ?? [];
    const currentDistributions = treeDistributionsByTimestamp.get(timestamp) ?? [];
    return {
      timestamp,
      harvests: currentHarvests.map((h) => ({
        tx: grabYieldGraphTx(h.id),
        timestamp,
        block: Number(h.blockNumber),
        token: vault.depositToken,
        amount: h.amount,
      })),
      treeDistributions: currentDistributions.map((d) => {
        const tokenAddress = d.token.id.startsWith('0x0x') ? d.token.id.slice(2) : d.token.id;
        return {
          tx: grabYieldGraphTx(d.id),
          timestamp,
          block: Number(d.blockNumber),
          token: ethers.utils.getAddress(tokenAddress),
          amount: d.amount,
        };
      }),
    };
  });
}

/**
 *
 * @param data
 * @returns
 */
function constructYieldItems(data: VaultHarvestData[]): VaultYieldItem[] {
  const recentHarvests = data.sort((a, b) => b.timestamp - a.timestamp);
  const allHarvests = recentHarvests.flatMap((h) => h.harvests.map((h) => ({ ...h, type: YieldType.Harvest })));
  const allDistributions = recentHarvests.flatMap((h) =>
    h.treeDistributions.map((d) => ({ ...d, type: YieldType.TreeDistribution })),
  );
  return allHarvests
    .concat(allDistributions)
    .filter((e) => BigNumber.from(e.amount).gt(ethers.constants.Zero))
    .sort((a, b) => b.timestamp - a.timestamp);
}

/**
 *
 * @param chain
 * @param vault
 * @param block
 * @returns
 */
async function getVaultBalance(chain: Chain, vault: VaultDefinitionModel, block: number): Promise<number> {
  const sdk = await chain.getSdk();

  let balance = 0;
  if (isInfluenceVault(vault.address)) {
    balance = await getInfuelnceVaultYieldBalance(chain, vault, block);
  } else {
    const vaultContract = Vault__factory.connect(vault.address, sdk.provider);
    const totalSupply = await vaultContract.totalSupply({ blockTag: block });
    balance = formatBalance(totalSupply);
  }

  return balance;
}

/**
 * Evaluate yield events to derivate base apr, and apy as well as apr of any emitted tokens.
 * @param chain network the vault is deployed on
 * @param vault vault requested yield event evaluation
 * @param yieldEvents events sourced from either on chain or the graph
 * @returns yield summary providing the base information for construction of yield sources
 */
async function evaluateYieldItems(
  chain: Chain,
  vault: VaultDefinitionModel,
  yieldItems: VaultYieldItem[],
): Promise<YieldEvent[]> {
  const yieldEvents: YieldEvent[] = [];
  const tokenEmissionAprs = new Map<string, number>();

  for (const item of yieldItems) {
    const block = item.block;
    const token = await getFullToken(chain, item.token);
    const amount = formatBalance(item.amount, token.decimals);
    const { price } = await queryPriceAtTimestamp(token.address, item.timestamp * 1000);

    const tokenEarned = price * amount;

    const balance = await getVaultBalance(chain, vault, item.block);
    const { price: vaultPrice } = await queryPriceAtTimestamp(vault.address, item.timestamp * 1000);
    const vaultPrincipal = vaultPrice * balance;
    const strategyInfo = await getStrategyInfo(chain, vault, { blockTag: item.block });
    const performanceScalar = 1 / (1 - strategyInfo.performanceFee / 10_000);

    const eventApr = calculateYield(vaultPrincipal, tokenEarned, VAULT_TWAY_DURATION);
    const yieldEvent: YieldEvent = {
      block,
      timestamp: item.timestamp * 1000,
      amount,
      token: token.address,
      type: item.type,
      value: vaultPrincipal,
      balance,
      earned: tokenEarned,
      apr: eventApr,
      grossApr: eventApr * performanceScalar,
      tx: item.tx,
    };
    yieldEvents.push(yieldEvent);

    if (item.type === YieldType.TreeDistribution) {
      const entry = tokenEmissionAprs.get(token.address) ?? 0;
      tokenEmissionAprs.set(token.address, entry + eventApr);
    }
  }

  return yieldEvents;
}

/**
 *
 * @param chain
 * @param vault
 * @param cutoff
 * @returns
 */
async function loadGraphYieldData(
  chain: Chain,
  vault: VaultDefinitionModel,
  cutoff: number,
): Promise<VaultHarvestData[]> {
  const { graph } = await chain.getSdk();
  const { address } = vault;
  console.log(`[${vault.name}]: Load Graph Yield Data`);
  const [vaultHarvests, treeDistributions] = await Promise.all([
    graph.loadSettHarvests({
      first: 25,
      where: {
        sett: address.toLowerCase(),
        timestamp_gt: cutoff,
      },
      orderBy: SettHarvest_OrderBy.Timestamp,
      orderDirection: OrderDirection.Asc,
    }),
    graph.loadBadgerTreeDistributions({
      first: 25,
      where: {
        sett: address.toLowerCase(),
        timestamp_gt: cutoff,
      },
      orderBy: BadgerTreeDistribution_OrderBy.Timestamp,
      orderDirection: OrderDirection.Asc,
    }),
  ]);
  const { settHarvests } = vaultHarvests;
  const { badgerTreeDistributions } = treeDistributions;
  return constructGraphVaultData(vault, settHarvests, badgerTreeDistributions);
}

/**
 *
 * @param chain
 * @param vault
 * @param lastHarvestBlock
 * @param cutoff
 * @returns
 */
async function loadEventYieldData(
  chain: Chain,
  vault: VaultDefinitionModel,
  lastHarvestBlock: number,
  cutoff: number,
): Promise<VaultHarvestData[]> {
  const { vaults } = await chain.getSdk();
  const { address, version } = vault;
  console.log(`[${vault.name}]: Load Event Yield Data`);
  try {
    const { data } = await vaults.listHarvests({
      address,
      timestamp_gte: cutoff,
      version,
      startBlock: lastHarvestBlock,
      endBlock: lastHarvestBlock + HARVEST_SCAN_BLOCK_INCREMENT,
    });
    return data;
  } catch {
    return loadGraphYieldData(chain, vault, cutoff);
  }
}

/**
 *
 * @param chain
 * @param vault
 * @param lastHarvestBlock
 * @returns
 */
export async function loadYieldEvents(
  chain: Chain,
  vault: VaultDefinitionModel,
  lastHarvestBlock: number,
): Promise<YieldEvent[]> {
  const { provider } = await chain.getSdk();
  const block = await provider.getBlock(lastHarvestBlock);
  const cutoff = block.timestamp;

  let data: VaultHarvestData[] = [];

  try {
    data = await loadGraphYieldData(chain, vault, cutoff);
  } catch (err) {
    if (isInfluenceVault(vault.address)) {
      throw err;
    } else {
      data = await loadEventYieldData(chain, vault, lastHarvestBlock, cutoff);
    }
  }

  const yieldItems = constructYieldItems(data);
  return evaluateYieldItems(chain, vault, yieldItems);
}

/**
 *
 * @param chain
 * @param vault
 * @returns
 */
export async function queryVaultYieldEvents(chain: Chain, vault: VaultDefinitionModel): Promise<VaultYieldEvent[]> {
  const mapper = getDataMapper();
  const chainAddress = getVaultEntityId(chain, vault);
  const cutoff = Date.now() - VAULT_TWAY_DURATION;

  const yieldEvents = [];
  for await (const yieldEvent of mapper.query(
    VaultYieldEvent,
    { chainAddress, timestamp: greaterThanOrEqualTo(cutoff) },
    { indexName: 'IndexYieldDataOnAddress', scanIndexForward: false },
  )) {
    yieldEvents.push(yieldEvent);
  }
  return yieldEvents;
}

/**
 *
 * @param chain
 * @param vault
 * @returns
 */
export async function queryVaultHistoricYieldEvents(
  chain: Chain,
  vault: VaultDefinitionModel,
): Promise<VaultYieldEvent[]> {
  // TODO: construct a new paginated 'all harvests' table
  // allow for page queries, timestamp based page queries
  // all pages will be the same size, new pages are only added after x entries to a given page
  // we can follow similar to the charting pattern with this data
  const mapper = getDataMapper();
  const chainAddress = getVaultEntityId(chain, vault);

  const yieldEvents = [];
  for await (const yieldEvent of mapper.query(
    VaultYieldEvent,
    { chainAddress },
    { indexName: 'IndexYieldDataOnAddress', scanIndexForward: false },
  )) {
    yieldEvents.push(yieldEvent);
  }
  return yieldEvents;
}
