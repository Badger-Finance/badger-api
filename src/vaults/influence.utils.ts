import { formatBalance, Vault__factory } from '@badger-dao/sdk';

import { Chain } from '../chains/config/chain.config';
import { TOKENS } from '../config/tokens.config';
import { getBalancerPoolTokens } from '../protocols/strategies/balancer.strategy';

// TODO: setup influence configs, voting periods, etc.
const influenceVaults = new Set([TOKENS.BVECVX, TOKENS.GRAVI_AURA]);

export function isInfluenceVault(address: string) {
  return influenceVaults.has(address);
}

export async function getInfuelnceVaultYieldBalance(
  chain: Chain,
  address: string,
  targetBlock: number,
): Promise<number> {
  const sdk = await chain.getSdk();
  const vaultContract = Vault__factory.connect(address, sdk.provider);
  const strategyBalance = await vaultContract.totalSupply({ blockTag: targetBlock });

  const maxBalance = formatBalance(strategyBalance);
  // TODO: tidy this up probably, or rethink with the configs how we deal with this
  if (address === TOKENS.BVECVX) {
    return maxBalance;
  }

  // deal with GRAVI_AURA - this won't scale so we need to figure out how to generalize it
  const excludedPools = [TOKENS.BPT_GRAV_AURABAL_WETH, TOKENS.BPT_GRAV_DIGG_WBTC];

  let blacklistedBalance = 0;
  for (const pool of excludedPools) {
    const tokens = await getBalancerPoolTokens(chain, pool);
    const graviAura = tokens.find((t) => t.address === TOKENS.GRAVI_AURA);
    if (graviAura) {
      blacklistedBalance += graviAura.balance;
    }
  }

  return maxBalance - blacklistedBalance;
}
