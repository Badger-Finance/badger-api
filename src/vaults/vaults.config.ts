import { ONE_DAY_SECONDS } from '@badger-dao/sdk';

export const VAULT_TWAY_PERIOD = 15;
export const VAULT_TWAY_DURATION = VAULT_TWAY_PERIOD * ONE_DAY_SECONDS;
export const VAULT_SOURCE = 'Vault Compounding';
