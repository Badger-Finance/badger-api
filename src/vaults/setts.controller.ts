import { Currency, Network } from '@badger-dao/sdk';
import { Controller, Get, Inject, PathParams, QueryParams } from '@tsed/common';
import { ContentType, Description, Returns, Summary } from '@tsed/schema';
import { Chain } from '../chains/config/chain.config';
import { VaultModel } from './interfaces/vault-model.interface';
import { VaultsService } from './vaults.service';
import { getVaultDefinition } from './vaults.utils';

@Controller('/setts')
export class SettsController {
  @Inject()
  settsService!: VaultsService;

  @Get()
  @ContentType('json')
  @Summary('Get a list of protocol setts')
  @Description('Return a list of protocol setts for the requested chain')
  @Returns(200, VaultModel)
  @Returns(400).Description('Not a valid chain')
  @Returns(404).Description('Not a valid sett')
  async listSetts(
    @QueryParams('chain') chain?: Network,
    @QueryParams('currency') currency?: Currency,
  ): Promise<VaultModel[]> {
    return this.settsService.listVaults(Chain.getChain(chain), currency);
  }

  @Get('/:vault')
  @ContentType('json')
  @Summary('Get a specific sett')
  @Description('Return a specific sett for the requested chain')
  @Returns(200, VaultModel)
  @Returns(400).Description('Not a valid chain')
  @Returns(404).Description('Not a valid sett')
  async getSett(
    @PathParams('contract') vault: string,
    @QueryParams('chain') chain?: Network,
    @QueryParams('currency') currency?: Currency,
  ): Promise<VaultModel> {
    return this.settsService.getVault(getVaultDefinition(Chain.getChain(chain), vault), currency);
  }
}
