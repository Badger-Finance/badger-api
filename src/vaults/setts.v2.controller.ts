import { Currency, Network } from '@badger-dao/sdk';
import { Controller, Get, Inject, PathParams, QueryParams } from '@tsed/common';
import { ContentType, Deprecated, Description, Returns, Summary } from '@tsed/schema';

import { Chain } from '../chains/config/chain.config';
import { VaultModel } from './interfaces/vault-model.interface';
import { VaultsService } from './vaults.service';
import { getVaultDefinition } from './vaults.utils';

@Deprecated()
@Controller('/setts')
export class SettsV2Controller {
  @Inject()
  settsService!: VaultsService;

  @Get()
  @ContentType('json')
  @Summary('Get a list of protocol setts')
  @Description('Return a list of protocol setts for the requested chain')
  @Returns(200, VaultModel)
  @Returns(400).Description('Not a valid chain')
  @Returns(404).Description('Not a valid sett')
  async listSetts(
    @QueryParams('chain') chain?: Network,
    @QueryParams('currency') currency?: Currency,
  ): Promise<VaultModel[]> {
    return this.settsService.listVaults(Chain.getChain(chain), currency);
  }

  @Get('/:vault')
  @ContentType('json')
  @Summary('Get a specific sett')
  @Description('Return a specific sett for the requested chain')
  @Returns(200, VaultModel)
  @Returns(400).Description('Not a valid chain')
  @Returns(404).Description('Not a valid sett')
  async getSett(
    @PathParams('vault') vault: string,
    @QueryParams('chain') chain?: Network,
    @QueryParams('currency') currency?: Currency,
  ): Promise<VaultModel> {
    const chainInst = Chain.getChain(chain);
    const vaultDef = getVaultDefinition(Chain.getChain(chain), vault);

    return this.settsService.getVault(chainInst, vaultDef, currency);
  }
}
