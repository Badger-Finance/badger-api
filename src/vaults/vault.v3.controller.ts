import { Currency, Network } from '@badger-dao/sdk';
import { Controller, Get, Inject, QueryParams, UseCache } from '@tsed/common';
import { ContentType, Description, Returns, Summary } from '@tsed/schema';
import { Chain } from '../chains/config/chain.config';
import { VaultModel } from './interfaces/vault-model.interface';
import { VaultsService } from './vaults.service';
import { getVaultDefinition } from './vaults.utils';
import { VaultHarvestsMapModel } from './interfaces/vault-harvests-list-model.interface';
import { VaultHarvestsMap } from './interfaces/vault-harvest-map';
import { VaultHarvestsModel } from './interfaces/vault-harvests-model.interface';
import { VaultHarvestsExtendedResp } from './interfaces/vault-harvest-extended-resp.interface';
import { QueryParamError } from '../errors/validation/query.param.error';

@Controller('/vault')
export class VaultsV3Controller {
  @Inject()
  vaultService!: VaultsService;

  @Get()
  @ContentType('json')
  @Summary('Get a specific vault')
  @Description('Return a specific vault for the requested chain')
  @Returns(200, VaultModel)
  @Returns(400).Description('Not a valid chain')
  @Returns(404).Description('Not a valid vault')
  async getVault(
    @QueryParams('vault') vault: string,
    @QueryParams('chain') chain?: Network,
    @QueryParams('currency') currency?: Currency,
  ): Promise<VaultModel> {
    if (!vault) throw new QueryParamError('vault');

    const chainInst = Chain.getChain(chain);
    const vaultDef = getVaultDefinition(Chain.getChain(chain), vault);

    return this.vaultService.getVault(chainInst, vaultDef, currency);
  }

  @Get('/list')
  @ContentType('json')
  @Summary('Get a list of protocol vaults')
  @Description('Return a list of protocol vaults for the requested chain')
  @Returns(200, VaultModel)
  @Returns(400).Description('Not a valid chain')
  async listVaults(
    @QueryParams('chain') chain?: Network,
    @QueryParams('currency') currency?: Currency,
  ): Promise<VaultModel[]> {
    return this.vaultService.listVaults(Chain.getChain(chain), currency);
  }

  @Get('/harvests')
  @ContentType('json')
  @Summary('Get harvests on a specific vault')
  @Description('Return full list of vault`s harvests')
  @Returns(200, Array).Of(VaultHarvestsModel)
  @Returns(400).Description('Not a valid chain')
  async getVaultsHarvests(
    @QueryParams('vault') vault: string,
    @QueryParams('chain') chain?: Network,
  ): Promise<VaultHarvestsExtendedResp[]> {
    if (!vault) throw new QueryParamError('vault');

    return this.vaultService.getVaultHarvests(Chain.getChain(chain), vault);
  }

  @Get('/list/harvests')
  @UseCache()
  @ContentType('json')
  @Summary('Get all vaults harvests on a chain')
  @Description('Return map of vaults, with harvests')
  @Returns(200, VaultHarvestsMapModel)
  @Returns(400).Description('Not a valid chain')
  async getlistVaultHarvests(@QueryParams('chain') chain?: Network): Promise<VaultHarvestsMap> {
    return this.vaultService.listVaultHarvests(Chain.getChain(chain));
  }
}
