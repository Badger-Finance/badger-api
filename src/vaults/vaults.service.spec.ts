import { Currency } from '@badger-dao/sdk';
import { PlatformTest } from '@tsed/common';

import { TOKENS } from '../config/tokens.config';
import * as pricesUtils from '../prices/prices.utils';
import { TEST_CHAIN } from '../test/tests.utils';
import { fullTokenMockMap } from '../tokens/mocks/full-token.mock';
import * as tokenUtils from '../tokens/tokens.utils';
import { VaultsService } from './vaults.service';
import * as vaultsUtils from './vaults.utils';

describe('proofs.service', () => {
  let service: VaultsService;

  beforeAll(async () => {
    await PlatformTest.create();
    service = PlatformTest.get<VaultsService>(VaultsService);
  });

  beforeEach(() => {
    jest.spyOn(tokenUtils, 'getFullToken').mockImplementation(async (_, tokenAddr) => {
      return fullTokenMockMap[tokenAddr] || fullTokenMockMap[TOKENS.BADGER];
    });
    jest.spyOn(vaultsUtils, 'getCachedVault').mockImplementation(async (chain, vault) => {
      const cachedVault = await vaultsUtils.defaultVault(chain, vault);
      cachedVault.value = Number(vault.address.slice(0, 6));
      cachedVault.balance = Number(vault.address.slice(0, 3));
      return cachedVault;
    });
    // TODO: implement pricing fixtures for test suites
    jest.spyOn(pricesUtils, 'convert').mockImplementation(async (price: number, currency?: Currency) => {
      if (currency === Currency.USD) {
        return price;
      }
      return (price * 8) / 3;
    });
  });

  afterEach(PlatformTest.reset);

  describe('getProtocolSummary', () => {
    describe('request with no currency', () => {
      it('returns the protocol summary in usd base currency', async () => {
        const result = await service.getProtocolSummary(TEST_CHAIN);
        expect(result).toMatchSnapshot();
      });
    });

    describe('request with currency', () => {
      it('returns the protocol summary in requested base currency', async () => {
        const result = await service.getProtocolSummary(TEST_CHAIN, Currency.AVAX);
        expect(result).toMatchSnapshot();
      });
    });
  });
});
