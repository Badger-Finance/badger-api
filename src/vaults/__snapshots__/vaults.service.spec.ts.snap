// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`proofs.service getProtocolSummary request with currency returns the protocol summary in requested base currency 1`] = `
Object {
  "setts": Array [
    Object {
      "balance": 13,
      "name": "renBTC/wBTC/sBTC",
      "value": 142197.33333333334,
    },
    Object {
      "balance": 6,
      "name": "renBTC/wBTC",
      "value": 75048,
    },
    Object {
      "balance": 11,
      "name": "tBTC/sBTC",
      "value": 126848,
    },
    Object {
      "balance": 10,
      "name": "renBTC/wBTC",
      "value": 119706.66666666667,
    },
    Object {
      "balance": 2,
      "name": "wBTC/Badger",
      "value": 24138.666666666668,
    },
    Object {
      "balance": 12,
      "name": "wBTC/Digg",
      "value": 132053.33333333334,
    },
    Object {
      "balance": 7,
      "name": "wBTC/wETH",
      "value": 80240,
    },
    Object {
      "balance": 1,
      "name": "wBTC/Badger",
      "value": 16645.333333333332,
    },
    Object {
      "balance": 8,
      "name": "wBTC/Digg",
      "value": 92890.66666666667,
    },
    Object {
      "balance": 7,
      "name": "Digg",
      "value": 86352,
    },
    Object {
      "balance": 1,
      "name": "Badger",
      "value": 17645.333333333332,
    },
    Object {
      "balance": 4,
      "name": "wBTC",
      "value": 51589.333333333336,
    },
    Object {
      "balance": 8,
      "name": "wBTC/ibBTC",
      "value": 94589.33333333333,
    },
    Object {
      "balance": 8,
      "name": "hBTC",
      "value": 95888,
    },
    Object {
      "balance": 5,
      "name": "pBTC",
      "value": 58413.333333333336,
    },
    Object {
      "balance": 15,
      "name": "oBTC",
      "value": 166082.66666666666,
    },
    Object {
      "balance": 5,
      "name": "bBTC",
      "value": 64037.333333333336,
    },
    Object {
      "balance": 11,
      "name": "Tricrypto",
      "value": 129728,
    },
    Object {
      "balance": 2,
      "name": "cvxCRV",
      "value": 29578.666666666668,
    },
    Object {
      "balance": 5,
      "name": "CVX",
      "value": 57194.666666666664,
    },
    Object {
      "balance": 2,
      "name": "Tricrypto2",
      "value": 27245.333333333332,
    },
    Object {
      "balance": 5,
      "name": "imBTC",
      "value": 61176,
    },
    Object {
      "balance": 2,
      "name": "mhBTC",
      "value": 26432,
    },
    Object {
      "balance": 15,
      "name": "bveCVX",
      "value": 172728,
    },
    Object {
      "balance": 9,
      "name": "CVX / bveCVX",
      "value": 100680,
    },
    Object {
      "balance": 10,
      "name": "ibBTC / crvsBTC",
      "value": 119184,
    },
    Object {
      "balance": 1,
      "name": "MIM / 3CRV",
      "value": 17674.666666666668,
    },
    Object {
      "balance": 1,
      "name": "FRAX / 3CRV",
      "value": 14877.333333333334,
    },
    Object {
      "balance": 6,
      "name": "remBadger",
      "value": 73021.33333333333,
    },
    Object {
      "balance": 14,
      "name": "BADGER / WBTC",
      "value": 161184,
    },
  ],
  "totalValue": 2435069.3333333335,
  "vaults": Array [
    Object {
      "balance": 13,
      "name": "renBTC/wBTC/sBTC",
      "value": 142197.33333333334,
    },
    Object {
      "balance": 6,
      "name": "renBTC/wBTC",
      "value": 75048,
    },
    Object {
      "balance": 11,
      "name": "tBTC/sBTC",
      "value": 126848,
    },
    Object {
      "balance": 10,
      "name": "renBTC/wBTC",
      "value": 119706.66666666667,
    },
    Object {
      "balance": 2,
      "name": "wBTC/Badger",
      "value": 24138.666666666668,
    },
    Object {
      "balance": 12,
      "name": "wBTC/Digg",
      "value": 132053.33333333334,
    },
    Object {
      "balance": 7,
      "name": "wBTC/wETH",
      "value": 80240,
    },
    Object {
      "balance": 1,
      "name": "wBTC/Badger",
      "value": 16645.333333333332,
    },
    Object {
      "balance": 8,
      "name": "wBTC/Digg",
      "value": 92890.66666666667,
    },
    Object {
      "balance": 7,
      "name": "Digg",
      "value": 86352,
    },
    Object {
      "balance": 1,
      "name": "Badger",
      "value": 17645.333333333332,
    },
    Object {
      "balance": 4,
      "name": "wBTC",
      "value": 51589.333333333336,
    },
    Object {
      "balance": 8,
      "name": "wBTC/ibBTC",
      "value": 94589.33333333333,
    },
    Object {
      "balance": 8,
      "name": "hBTC",
      "value": 95888,
    },
    Object {
      "balance": 5,
      "name": "pBTC",
      "value": 58413.333333333336,
    },
    Object {
      "balance": 15,
      "name": "oBTC",
      "value": 166082.66666666666,
    },
    Object {
      "balance": 5,
      "name": "bBTC",
      "value": 64037.333333333336,
    },
    Object {
      "balance": 11,
      "name": "Tricrypto",
      "value": 129728,
    },
    Object {
      "balance": 2,
      "name": "cvxCRV",
      "value": 29578.666666666668,
    },
    Object {
      "balance": 5,
      "name": "CVX",
      "value": 57194.666666666664,
    },
    Object {
      "balance": 2,
      "name": "Tricrypto2",
      "value": 27245.333333333332,
    },
    Object {
      "balance": 5,
      "name": "imBTC",
      "value": 61176,
    },
    Object {
      "balance": 2,
      "name": "mhBTC",
      "value": 26432,
    },
    Object {
      "balance": 15,
      "name": "bveCVX",
      "value": 172728,
    },
    Object {
      "balance": 9,
      "name": "CVX / bveCVX",
      "value": 100680,
    },
    Object {
      "balance": 10,
      "name": "ibBTC / crvsBTC",
      "value": 119184,
    },
    Object {
      "balance": 1,
      "name": "MIM / 3CRV",
      "value": 17674.666666666668,
    },
    Object {
      "balance": 1,
      "name": "FRAX / 3CRV",
      "value": 14877.333333333334,
    },
    Object {
      "balance": 6,
      "name": "remBadger",
      "value": 73021.33333333333,
    },
    Object {
      "balance": 14,
      "name": "BADGER / WBTC",
      "value": 161184,
    },
  ],
}
`;

exports[`proofs.service getProtocolSummary request with no currency returns the protocol summary in usd base currency 1`] = `
Object {
  "setts": Array [
    Object {
      "balance": 13,
      "name": "renBTC/wBTC/sBTC",
      "value": 142197.33333333334,
    },
    Object {
      "balance": 6,
      "name": "renBTC/wBTC",
      "value": 75048,
    },
    Object {
      "balance": 11,
      "name": "tBTC/sBTC",
      "value": 126848,
    },
    Object {
      "balance": 10,
      "name": "renBTC/wBTC",
      "value": 119706.66666666667,
    },
    Object {
      "balance": 2,
      "name": "wBTC/Badger",
      "value": 24138.666666666668,
    },
    Object {
      "balance": 12,
      "name": "wBTC/Digg",
      "value": 132053.33333333334,
    },
    Object {
      "balance": 7,
      "name": "wBTC/wETH",
      "value": 80240,
    },
    Object {
      "balance": 1,
      "name": "wBTC/Badger",
      "value": 16645.333333333332,
    },
    Object {
      "balance": 8,
      "name": "wBTC/Digg",
      "value": 92890.66666666667,
    },
    Object {
      "balance": 7,
      "name": "Digg",
      "value": 86352,
    },
    Object {
      "balance": 1,
      "name": "Badger",
      "value": 17645.333333333332,
    },
    Object {
      "balance": 4,
      "name": "wBTC",
      "value": 51589.333333333336,
    },
    Object {
      "balance": 8,
      "name": "wBTC/ibBTC",
      "value": 94589.33333333333,
    },
    Object {
      "balance": 8,
      "name": "hBTC",
      "value": 95888,
    },
    Object {
      "balance": 5,
      "name": "pBTC",
      "value": 58413.333333333336,
    },
    Object {
      "balance": 15,
      "name": "oBTC",
      "value": 166082.66666666666,
    },
    Object {
      "balance": 5,
      "name": "bBTC",
      "value": 64037.333333333336,
    },
    Object {
      "balance": 11,
      "name": "Tricrypto",
      "value": 129728,
    },
    Object {
      "balance": 2,
      "name": "cvxCRV",
      "value": 29578.666666666668,
    },
    Object {
      "balance": 5,
      "name": "CVX",
      "value": 57194.666666666664,
    },
    Object {
      "balance": 2,
      "name": "Tricrypto2",
      "value": 27245.333333333332,
    },
    Object {
      "balance": 5,
      "name": "imBTC",
      "value": 61176,
    },
    Object {
      "balance": 2,
      "name": "mhBTC",
      "value": 26432,
    },
    Object {
      "balance": 15,
      "name": "bveCVX",
      "value": 172728,
    },
    Object {
      "balance": 9,
      "name": "CVX / bveCVX",
      "value": 100680,
    },
    Object {
      "balance": 10,
      "name": "ibBTC / crvsBTC",
      "value": 119184,
    },
    Object {
      "balance": 1,
      "name": "MIM / 3CRV",
      "value": 17674.666666666668,
    },
    Object {
      "balance": 1,
      "name": "FRAX / 3CRV",
      "value": 14877.333333333334,
    },
    Object {
      "balance": 6,
      "name": "remBadger",
      "value": 73021.33333333333,
    },
    Object {
      "balance": 14,
      "name": "BADGER / WBTC",
      "value": 161184,
    },
  ],
  "totalValue": 2435069.3333333335,
  "vaults": Array [
    Object {
      "balance": 13,
      "name": "renBTC/wBTC/sBTC",
      "value": 142197.33333333334,
    },
    Object {
      "balance": 6,
      "name": "renBTC/wBTC",
      "value": 75048,
    },
    Object {
      "balance": 11,
      "name": "tBTC/sBTC",
      "value": 126848,
    },
    Object {
      "balance": 10,
      "name": "renBTC/wBTC",
      "value": 119706.66666666667,
    },
    Object {
      "balance": 2,
      "name": "wBTC/Badger",
      "value": 24138.666666666668,
    },
    Object {
      "balance": 12,
      "name": "wBTC/Digg",
      "value": 132053.33333333334,
    },
    Object {
      "balance": 7,
      "name": "wBTC/wETH",
      "value": 80240,
    },
    Object {
      "balance": 1,
      "name": "wBTC/Badger",
      "value": 16645.333333333332,
    },
    Object {
      "balance": 8,
      "name": "wBTC/Digg",
      "value": 92890.66666666667,
    },
    Object {
      "balance": 7,
      "name": "Digg",
      "value": 86352,
    },
    Object {
      "balance": 1,
      "name": "Badger",
      "value": 17645.333333333332,
    },
    Object {
      "balance": 4,
      "name": "wBTC",
      "value": 51589.333333333336,
    },
    Object {
      "balance": 8,
      "name": "wBTC/ibBTC",
      "value": 94589.33333333333,
    },
    Object {
      "balance": 8,
      "name": "hBTC",
      "value": 95888,
    },
    Object {
      "balance": 5,
      "name": "pBTC",
      "value": 58413.333333333336,
    },
    Object {
      "balance": 15,
      "name": "oBTC",
      "value": 166082.66666666666,
    },
    Object {
      "balance": 5,
      "name": "bBTC",
      "value": 64037.333333333336,
    },
    Object {
      "balance": 11,
      "name": "Tricrypto",
      "value": 129728,
    },
    Object {
      "balance": 2,
      "name": "cvxCRV",
      "value": 29578.666666666668,
    },
    Object {
      "balance": 5,
      "name": "CVX",
      "value": 57194.666666666664,
    },
    Object {
      "balance": 2,
      "name": "Tricrypto2",
      "value": 27245.333333333332,
    },
    Object {
      "balance": 5,
      "name": "imBTC",
      "value": 61176,
    },
    Object {
      "balance": 2,
      "name": "mhBTC",
      "value": 26432,
    },
    Object {
      "balance": 15,
      "name": "bveCVX",
      "value": 172728,
    },
    Object {
      "balance": 9,
      "name": "CVX / bveCVX",
      "value": 100680,
    },
    Object {
      "balance": 10,
      "name": "ibBTC / crvsBTC",
      "value": 119184,
    },
    Object {
      "balance": 1,
      "name": "MIM / 3CRV",
      "value": 17674.666666666668,
    },
    Object {
      "balance": 1,
      "name": "FRAX / 3CRV",
      "value": 14877.333333333334,
    },
    Object {
      "balance": 6,
      "name": "remBadger",
      "value": 73021.33333333333,
    },
    Object {
      "balance": 14,
      "name": "BADGER / WBTC",
      "value": 161184,
    },
  ],
}
`;
